### PR TITLE
[DynamoDb] Adding BatchGetItem

### PR DESCRIPTION
**Pull request by Nyholm**
After @Nemo64's work on recursion, I though I could give this a try. 

There are some issues currently with the generator. I'll look into fix them. 

Related to #566

### CODE REVIEW

File: src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php [hunk 1]
```diff
@@ -267,9 +267,9 @@ private function parseResponseMap(MapShape $shape, string $input, bool $required
                     $this->functions[$keyName] = strtr($body, [
                         'FUNCTION_KEY' => \var_export($keyName, true),
                         'CLASS' => $shape->getValue()->getShape()->getName(),
-                    ]);
+                    ]);// TODO add CLASS to imports
```
Comment (Nyholm) [inline src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php:270]: How can I do this in a good way?

File: src/Service/CognitoIdentityProvider/src/Result/AdminInitiateAuthResponse.php [hunk 2]
```diff
@@ -78,7 +78,7 @@ protected function populateResult(Response $response): void
 
         $this->ChallengeName = isset($data['ChallengeName']) ? (string) $data['ChallengeName'] : null;
         $this->Session = isset($data['Session']) ? (string) $data['Session'] : null;
-        $this->ChallengeParameters = empty($data['ChallengeParameters']) ? [] : (function (array $json): array {
+        $this->ChallengeParameters = empty($data['ChallengeParameters']) ? [] : (function (array $json) use (&$fn): array {
```
Comment (Nyholm) [inline src/Service/CognitoIdentityProvider/src/Result/AdminInitiateAuthResponse.php:81]: This is a side effect. It is not pretty. Can we care?

File: src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php (hunk 1 above)
Comment (jderusse) [reply src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php]: I hesitate between 
- returning the list of classes used (and to register)
- passing the namespace to the method to let the parser/dumper register the class by themselves
- creating a ClassUsageService injected in all code generators with 3 methods `start($namespace)`, `end()`, `use(ClassName $class)`

File: src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php [hunk 3]
```diff
@@ -267,9 +267,9 @@ private function parseResponseMap(MapShape $shape, string $input, bool $required
                     $this->functions[$keyName] = strtr($body, [
                         'FUNCTION_KEY' => \var_export($keyName, true),
                         'CLASS' => $shape->getValue()->getShape()->getName(),
-                    ]);
+                    ]); // TODO add CLASS to imports
```
Comment (Nyholm) [inline src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php:270]: How can I add `CLASS` as a `use` statement? It is always a value object.

File: src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php (hunk 1 above)
Comment (Nyholm) [reply src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php]: There is a forth option: Just add all value objects as `use` statements and let cs-fixer remove them. 

Im not sure what the best way forward is. I'll explore your options first.

Comment (Nyholm) [reply src/CodeGenerator/src/Generator/ResponseParser/RestJsonParser.php]: I returned a list of classes. https://github.com/async-aws/aws/pull/677/commits/55dba4105f55b39a7f9f3843b36b2daeab819a98

File: src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\CodeGenerator\Generator\ResponseParser;
+
+/**
+ * @internal
+ */
+class ParserResult
```
Comment (jderusse) [inline src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php:10]: I like this new class!

File: src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        $this->body = $body;
+        $this->imports = $imports;
+    }
+
+    public function getBody(): string
+    {
+        return $this->body;
+    }
+
+    public function getImports(): array
```
Comment (jderusse) [inline src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php:35]: `getUsedClasses` ?

File: src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace AsyncAws\CodeGenerator\Generator\ResponseParser;
+
+/**
+ * @internal
+ */
+class ParserResult
+{
+    /**
+     * @var string
```
Comment (jderusse) [inline src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php:13]: not needed

File: src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php (hunk 4 above)
Comment (Nyholm) [reply src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php:10]: Thank you

File: src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php (hunk 5 above)
Comment (Nyholm) [reply src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php]: Updated.

File: src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php (hunk 6 above)
Comment (Nyholm) [reply src/CodeGenerator/src/Generator/ResponseParser/ParserResult.php]: Removed

**Comment by Nyholm** [comment]
Im ready for a review

**Pull request merged**